File: src/main/java/com/trie/ecommerce/entity/Product.java
package com.trie.ecommerce.entity;

import com.trie.ecommerce.enums.PricingType;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "products")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank
    @Column(nullable = false)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @NotBlank
    @Column(nullable = false)
    private String category;

    private String material;

    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PricingType pricingType;

    @Column(precision = 10, scale = 2)
    private BigDecimal pricePerGram;

    @Builder.Default
    private Boolean active = true;

    @Column(updatable = false)
    private LocalDateTime createdAt;

    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<ProductVariant> variants = new ArrayList<>();

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: front/src/types/product.ts
export type PricingType = 'FIXED' | 'BY_GRAM';

export interface Variant {
  id: number;
  size: string | null;
  weightInGrams: number | null;
  price: number;
  sku: string;
  isUniquePiece: boolean;
  stockQuantity: number | null;
}

export interface Product {
  id: number;
  name: string;
  description: string | null;
  category: string | null;
  material: string | null;
  pricingType: PricingType;
  pricePerGram: number | null;
  active: boolean;
  imageUrl: string | null;
  featured: boolean;
  newCollection: boolean;
  createdAt: string;
  variants: Variant[];
}
